galsim/phase_screens.py

```diff
@@ -156,14 +258,72 @@ def __init__(self, screen_size, screen_scale=None, altitude=0.0, r0_500=0.2, L0=
         self.dynamic = True
         self.reversible = self.alpha == 1.0
 
-        # These will be None until screens are instantiated.
-        self.kmin = None
-        self.kmax = None
+        # Use shared memory for screens.  Allocate it here; fill it in on demand.
+        # Shared memory implementation depends on python version.
+        if sys.version_info >= (3,4):
+            global _GSScreenShare
+            if isinstance(mp_context, multiprocessing.context.BaseContext):
+                ctx = mp_context
+            else:
+                ctx = multiprocessing.get_context(mp_context)
+            self._objDict = {
+                'f':ctx.RawArray('d', (self.npix+1)*(self.npix+1)),
+                'x':ctx.RawArray('d', self.npix+1),
+                'y':ctx.RawArray('d', self.npix+1),
+                'kmin':ctx.RawValue('d', -1.0),
+                'kmax':ctx.RawValue('d', -1.0),
+                'x0':ctx.RawValue('d', 0.0),
+                'y0':ctx.RawValue('d', 0.0),
+                'xperiod':ctx.RawValue('d', 0.0),
+                'yperiod':ctx.RawValue('d', 0.0),
+                'instantiated':ctx.RawValue('b', False),
+                'refcount':ctx.RawValue('i', 1),
+                'lock':ctx.Lock()
+            }
+            # A unique id for this screen, created in the parent process, that can be used to find the
+            # correct shared memory object in child processes.
+            self._shareKey = id(self)
+            _GSScreenShare[self._shareKey] = self._objDict
+        else:
+            self._kmin = None
+            self._kmax = None
+            self._instantiated = False
+
+    def __setstate__(self, d):
+        self.__dict__.update(d)
+        if sys.version_info >= (3,4):
+            self._objDict = _GSScreenShare[self._shareKey]
+            with self._objDict['lock']:
+                self._objDict['refcount'].value += 1
+
+    def __getstate__(self):
+        d = self.__dict__.copy()
+        if sys.version_info >= (3,4):
+            d.pop('_objDict')
+        return d
+
+    def __del__(self):
+        if sys.version_info >= (3,4):
+            if not hasattr(self, '_objDict'):  # for if __init__ raised exception
+                return
+
+            with self._objDict['lock']:
+                self._objDict['refcount'].value -= 1
+            if self._objDict['refcount'].value == 0:
+                del _GSScreenShare[self._shareKey]
```

Comment:
I feel like this del bit should also be inside the lock.  Isn't there a race potential for the refcount to be 0, but someone else is about to increment it?

Reply:
Right, I think moving the if and del inside the lock saves you from deallocating in process A while there's a copy happening in process B.  Maybe even use additional `with lock` blocks to give a potential copy an extra chance to increment the refcount.  If the del continues to win a del/copy race, the copy will still fail I guess, but without corrupting memory.  It's probably not possible to ensure that in a copy/delete race the copy always wins.  We could move the del to a garbage collect function, but that seems like more work and more user-responsibility than reasonable.

Reply:
I think this is now reasonably race-condition resistant.